src/cli/serve/integration_tests/config_ordering.test.ts

```diff
@@ -0,0 +1,192 @@
+import * as Os from 'os';
+import * as Child from 'child_process';
+import Del from 'del';
+import * as Rx from 'rxjs';
+import { filter, map, take, timeout } from 'rxjs/operators';
+
+const tempDir = Path.join(Os.tmpdir(), 'kbn-config-test');
+
+const kibanaPath = follow('../../../../scripts/kibana.js');
+const configsPath = follow('../../../../config');
+
+const TestFiles = {
+  fileList: [] as string[],
+
+  createEmptyConfigFiles(fileNames: string[], root: string = tempDir): string[] {
+    const configFiles = [];
+    for (const fileName of fileNames) {
+      const filePath = Path.resolve(root, fileName);
+
+      if (!Fs.existsSync(filePath)) {
+        Fs.writeFileSync(filePath, 'dummy');
+
+        TestFiles.fileList.push(filePath);
+      }
+
+      configFiles.push(filePath);
+    }
+
+    return configFiles;
+  },
+  cleanUpEmptyConfigFiles() {
+    for (const filePath of TestFiles.fileList) {
+      Del.sync(filePath);
+    }
+    TestFiles.fileList.length = 0;
+  },
+};
+
+describe('Server configuration ordering', () => {
+  let kibanaProcess: Child.ChildProcessWithoutNullStreams;
+
+  beforeEach(() => {
+    Fs.mkdirSync(tempDir, { recursive: true });
+  });
+
+  afterEach(async () => {
+    if (kibanaProcess !== undefined) {
+      const exitPromise = new Promise((resolve) => kibanaProcess?.once('exit', resolve));
+      kibanaProcess.kill('SIGKILL');
+      await exitPromise;
+    }
+
+    Del.sync(tempDir, { force: true });
+    TestFiles.cleanUpEmptyConfigFiles();
+  });
+
+  it('loads default config set without any options', async function () {
+    kibanaProcess = Child.spawn(process.execPath, [kibanaPath]);
+
+    const configList = await extractConfigurationOrder(kibanaProcess);
+
+    expect(configList).toEqual(['kibana.yml']);
+  });
+
+  it('loads serverless configs when --serverless is set', async () => {
+    kibanaProcess = Child.spawn(process.execPath, [kibanaPath, '--serverless', 'oblt']);
+
+    const configList = await extractConfigurationOrder(kibanaProcess);
+
+    expect(configList).toEqual([
+      'serverless.yml',
+      'serverless.oblt.yml',
+      'kibana.yml',
+      'serverless.recent.yml',
+    ]);
+  });
+
+  it('prefers --config options over default', async () => {
+    const [configPath] = TestFiles.createEmptyConfigFiles(['potato.yml']);
+
+    kibanaProcess = Child.spawn(process.execPath, [
+      kibanaPath,
+      '--serverless',
+      'oblt',
+      '--config',
+      configPath,
+    ]);
+
+    const configList = await extractConfigurationOrder(kibanaProcess);
+
+    expect(configList).toEqual([
+      'serverless.yml',
+      'serverless.oblt.yml',
+      'potato.yml',
+      'serverless.recent.yml',
+    ]);
+  });
+
+  it('adds dev configs to the queue', async () => {
+    TestFiles.createEmptyConfigFiles(
+      [
+        'kibana.dev.yml',
+        'serverless.dev.yml',
+        // 'serverless.es.dev.yml' // Shouldn't this be loaded? It's mentioned in the README, but wasn't in the code
```

Comment:
We've typically left the file uncommitted and untracked as a way for developers to persist setups.  Optionally loaded.